Add raffle contract

eloylp opened the pull request:
This adds a basic raffle example with following features:

- Implemented the token interface, so it should be able to transfer funds in a real environment.
- It can work with any token that accomplishes the token interface.
- The number of winners is customizable.
- The cost per ticket is customizable.
- Admin should introduce random seed when playing the raffle, that will be combined with ledger unix time.
- It calculates the balances before transfering ticket amount.
- Only admin can play the raffle, so triggering the funds spread among the winners.
- Prevents playing a raffle with no participants.

eloylp commented on contracts/lottery/src/lib.rs:
```diff
@@ -0,0 +1,136 @@
+        storage.set(&DataKey::WinnerCount, &winners_count);
+        storage.set(&DataKey::TicketPrice, &ticket_price);
+        storage.set(&DataKey::Candidates, &Vec::<Address>::new(&env));
+        storage.set(&DataKey::AlreadyPlayed, &false);
```
Probably in the wizard we would remove the init parameters, and "hardcode" them by substitution.

geofmureithi replied:
These would be passed in invoke

eloylp replied:
Yes, they already are, this comment is from 3 weeks ago. I will resolve it to not cause confusion.

mariopil commented on contracts/lottery/src/test.rs:
```diff
@@ -0,0 +1,201 @@
+    client.initialize(&client.address, &test_token_client.address, &1, &100);
+
+    client.play_raffle(&666);
+}
```
Missing new line.

eloylp replied:
fixed at 292039c3979c00397c437659072e1d4b8414cba5

eloylp commented on contracts/lottery/src/lib.rs:
```diff
@@ -0,0 +1,136 @@
+        if winners.contains_key(winner) {
+            continue;
+        }
+        winners.set(winner, true);
```
Its a great question. Originally i tried to fix the "duplicated winners" problem there, but that means we are probably not going to achieve the exact numbers of winners as you said.

For now, i would probably rename the variable to `max_winners_count` and add a comment in code regarding this. That way,  we avoid keep looping undefinetly for finding winners, as looks execution time is [constrained to seconds](https://soroban.stellar.org/docs/fundamentals-and-concepts/fees-and-metering#resource-limits). That doesn't mean we are not going to better improve this contract later :)

Wdyt ? @mariopil @geofmureithi

eloylp replied:
I did this guys 2f960d912895de57c0ca91f53a26603c07e9f1be let me know your thoughts.

eloylp commented:
@mariopil nomenclature problem was fixed at 511ee87f471bb92b296ad5d79c1c8e7de34d4fb2

geofmureithi commented on README.md:
I wish this was in a different PR

eloylp replied:
Yeah, initially, this PR was created before all the work we did here #16 . I think it would be good to start merging things. I would start merging #16 and then i will take care of solving conflicts here. Wdyt ?

eloylp replied:
@geofmureithi , i just reverted the commits, so theres no changes in the readme for this PR. 

All Readme.md changes are hosted in #16 , feel free to review it.

eloylp commented:
Hello @geofmureithi . Some changes here:
* Added the rust doc comments to this contract. 
* Added some extra security checkings (see commits from today).
* Eliminated all "Lottery" nomenclature, as we were mixing raffle and lottery. The difference is in raffle there is always winner, which is our current case.  
* I also updated the Raffle contract, but not the docs, as that breaks JS.  In a future, i think we should look how to directly read contracts from repo in the wizard and solve everything in one shoot : )

Requesting final review !

eloylp commented on app/src/contractparams/VotingParams.tsx:
```diff
@@ -25,7 +25,7 @@ export class VotingParams extends React.Component<Props> {
     return (
       <form>
-        <div className="LotteryParams">
+        <div className="RaffleParams">
```
No i do not think so. Checked the UI with `npm start`  and it also not appear referenced in any css.
